### PR TITLE
feat: Expression extension framework

### PR DESCRIPTION
**Pull request by valya**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,5 +54,10 @@
     "ejs": "^3.1.8",
     "fork-ts-checker-webpack-plugin": "^6.0.4",
     "globby": "^11.0.2"
+  },
+  "dependencies": {
```
Comment (netroy) [inline package.json:58]: these can be removed, right?
These seem to already be defined in `packages/workflow/package.json`

File: packages/workflow/src/Expression.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
+// @ts-ignore
```
Comment (netroy) [inline packages/workflow/src/Expression.ts:2]: we don't need to `@ts-ignore`. we have explicit typings for '@n8n_io/riot-tmpl'

File: packages/workflow/src/Expression.ts [hunk 3]
```diff
@@ -14,14 +16,28 @@ import {
...
+import { extend, hasExpressionExtension, hasNativeMethod } from './Extensions';
+import {
+	ExpressionChunk,
+	ExpressionCode,
+	joinExpression,
+	splitExpression,
+} from './Extensions/ExpressionParser';
+import { extendTransform } from './Extensions/ExpressionExtension';
+
+// @ts-ignore
```
Comment (netroy) [inline packages/workflow/src/Expression.ts:33]: unused

File: packages/workflow/src/Expression.ts [hunk 4]
```diff
@@ -14,14 +16,28 @@ import {
...
+
 // Set it to use double curly brackets instead of single ones
+// eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
```
Comment (netroy) [inline packages/workflow/src/Expression.ts:36]: not needed

File: packages/workflow/src/Expression.ts [hunk 5]
```diff
@@ -14,14 +16,28 @@ import {
...
 tmpl.brackets.set('{{ }}');
 
 // Make sure that error get forwarded
+// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
```
Comment (netroy) [inline packages/workflow/src/Expression.ts:40]: not needed

File: packages/workflow/src/Expression.ts [hunk 6]
```diff
@@ -266,8 +286,12 @@ export class Expression {
 		return returnValue;
 	}
 
-	private renderExpression(expression: string, data: IWorkflowDataProxyData): tmpl.ReturnValue {
+	private renderExpression(
+		expression: string,
+		data: IWorkflowDataProxyData,
+	): tmpl.ReturnValue | undefined {
 		try {
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
```
Comment (netroy) [inline packages/workflow/src/Expression.ts:294]: not needed

File: package.json (hunk 1 above)
Comment (valya) [reply package.json:58]: Ahh, yeah. This likely happened in one of the 5 times I had to rebase to remove Brian's email.

**Comment by netroy** [comment]
I don't know enough about our current expressions code to approve, but overall (besides the minor linting issues) this looks good to me. Thanks for adding all the tests.

File: packages/workflow/src/Extensions/StringExtensions.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+			.replace(/\n{2,}/g, '\n\n');
+	} catch (e) {
+		return value;
+	}
+	return output;
+}
+
+function sayHi(value: string) {
+	return `hi ${value}`;
+}
```
Comment (krynble) [inline packages/workflow/src/Extensions/StringExtensions.ts:106]: We can remove this function since it was used just for testing

**Review by krynble — CHANGES_REQUESTED** [review]
It's not something severe but I was able to break the expression parser by using the following expression:

`{{$json["propertyName"].sayHi().substring(4).substring(2) + '{{' + "}}" }}`

File: packages/workflow/src/Extensions/DateExtensions.ts [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+	secs: 'second',
+	sec: 'second',
+	ms: 'millisecond',
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isDateTime(date: any): date is DateTime {
+	if (date) {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+		return DateTime.isDateTime(date);
```
[the remark this answers is not in the document]
Comment (valya) [reply packages/workflow/src/Extensions/DateExtensions.ts:74]: This is an internal function for this file only. It's used to check if it's a `DateTime` as opposed to a `Date` object. Anything that comes into the date functions will either be a `Date` or `DateTime` due to [this check](https://github.com/n8n-io/n8n/blob/5a8b553c471ebf6470f2bc52aabbc10d9d4f8701/packages/workflow/src/Extensions/ExpressionExtension.ts#L158).

**Review by krynble — DISMISSED** [review]

**Review by krynble — APPROVED** [review]

File: packages/workflow/test/ExpressionExtensions/DateExtensions.test.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+			expect(evaluate('={{ DateTime.now().format("yyyy LLL dd") }}')).toEqual(
+				dateExtensions.functions.format(new Date(), ['yyyy LLL dd']),
+			);
+			expect(evaluate('={{ DateTime.now().format("yyyy LLL dd") }}')).not.toEqual(
+				dateExtensions.functions.format(new Date(), ["HH 'hours and' mm 'minutes'"]),
+			);
+		});
+
+		test('.toDate() should work on a string', () => {
+			expect(evaluate('={{ "2022-01-03T00:00:00.000+00:00".toDate() }}')).toEqual(new Date(2022, 0, 3));
```
Comment (netroy) [inline packages/workflow/test/ExpressionExtensions/DateExtensions.test.ts:50]: This test is consistently failing for me locally. Looks like a UTC/Timezone issue 
image

Comment (netroy) [reply packages/workflow/test/ExpressionExtensions/DateExtensions.test.ts:50]: looks like this test will only pass on a system running in UTC, as `new Date(2022, 0, 3)` will produce a different ISO string for different timezomes.

**Comment by janober** [comment]
Got released with `n8n@0.211.0`

**Pull request merged**